src/config_store.rs

```diff
@@ -0,0 +1,169 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+extern crate serde_json;
+
+use std::collections::BTreeMap;
+use std::fs;
+use std::fs::File;
+use std::io::prelude::*;
+use std::path::Path;
+use std::sync::Mutex;
+
+
+type ConfigNameSpace = BTreeMap<String, String>;
+
+type ConfigTree = BTreeMap<String, ConfigNameSpace>;
+
+#[derive(Debug)]
+pub struct ConfigStore {
+    file_name: String,
+    save_lock: Mutex<()>,
+    config: ConfigTree
+}
+
+impl ConfigStore {
+    pub fn new(file_name: String) -> Self {
+        ConfigStore {
+            file_name: file_name.clone(),
+            save_lock: Mutex::new(()),
+            config: ConfigStore::load(file_name)
+        }
+    }
+
+    pub fn set(&mut self, namespace: String, property: String, value: String) {
+        debug!("Setting config for {}::{} to {}", namespace, property, value);
+        if !self.config.contains_key(&namespace) {
+            self.config.insert(namespace.clone(), ConfigNameSpace::new());
+        }
+        self.config.get_mut(&namespace).unwrap().insert(property, value);
+        // TODO: Should be more intelligent than save on every write
+        self.save();
+    }
+
+    pub fn get(&self, namespace: String, property: String) -> Option<&String> {
+        if self.config.contains_key(&namespace) {
+            let res = self.config.get(&namespace).unwrap().get(&property);
+            debug!("Config result for {}::{} is {:?}", namespace, property, res);
+            res
+        } else {
+            debug!("No config result for {}::{}", namespace, property);
+            None
+        }
+    }
+
+    fn load(file_name: String) -> ConfigTree {
+        let empty_config = BTreeMap::new();
+        let path = Path::new(&file_name);
+        let file = match File::open(&path) {
+            Ok(file) => {
+                file
+            },
+            Err(error) => {
+                error!("Unable to open configuration file {}: {}",
+                    file_name, error.to_string());
+                return empty_config;
+            }
+        };
+        let parsed_config: ConfigTree = match serde_json::from_reader(&file) {
+            Ok(value) => value,
+            Err(error) => {
+                error!("Unable to generate JSON from config file {}: {}",
+                    file_name, error.to_string());
+                return empty_config;
+            }
+        };
+
+        debug!("Parsed config file: {:?}", parsed_config);
+        parsed_config
+    }
+
+    #[allow(unused_variables)]
```

Comment:
Is that for lock = ... ? If so, replace by let _ = self.save_lock.lock().unwrap()